File: AddOperators.java
// Time Complexity : O(N3^N) where N is length of input string
// Space Complexity : O(3^N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class AddOperators {
    class Solution {
        public List<String> addOperators(String num, int target)
        {
            List<String> res = new ArrayList<>();
            StringBuilder sb = new StringBuilder();
            dfs(res, sb, num, 0, target, 0, 0);
            return res;

        }
        public void dfs(List<String> res, StringBuilder sb, String num, int pos, int target, long prev, long multi)
        {
            if(pos == num.length()) {
                if(target == prev) res.add(sb.toString());
                return;
            }
            for(int i = pos; i < num.length(); i++) {
                if(num.charAt(pos) == '0' && i != pos) break;
                long curr = Long.parseLong(num.substring(pos, i + 1));
                int len = sb.length();
                if(pos == 0) {
                    dfs(res, sb.append(curr), num, i + 1, target, curr, curr);
                    sb.setLength(len);
                } else {
                    dfs(res, sb.append("+").append(curr), num, i + 1, target, prev + curr, curr);
                    sb.setLength(len);
                    dfs(res, sb.append("-").append(curr), num, i + 1, target, prev - curr, -curr);
                    sb.setLength(len);
                    dfs(res, sb.append("*").append(curr), num, i + 1, target, prev - multi + multi * curr, multi * curr);
                    sb.setLength(len);
                }
            }
        }
    }
}

File: addOperators.js
// Time Complexity : O(n3^n) where n is num length
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes, but when submitting one input is not working.
// Any problem you faced while coding this : Yes, input (Input : "3456237490" 9191) is not working and I am not able to figure out what is wrong in the code. Other inputs seems to be working.

/**
 * @param {string} num
 * @param {number} target
 * @return {string[]}
 */
let result;
var helper = function(num, index, calc, tail, path, target) {
    //base
    if(index === num.length) {
        if(calc === target) {
            result.push(path);
        }
        return;
    }
    
    //logic
    for(let i = index; i < num.length; i++) {
        //preceding zero
        if(num.charAt(index) == '0' && index != i) continue; 
        let curr = Number(num.substring(index, i+1));
        if(index === 0) {
            //action
            helper(num, i + 1, curr, curr, path + curr, target)
        } else {
            //+
            helper(num, i + 1, calc + curr, curr, path + "+" + curr, target)
            //-
            helper(num, i + 1, calc - curr, curr, path + "-" + curr, target)
            //*
            helper(num, i + 1, calc - tail + tail * curr, tail * curr, path + "*" + curr, target)
        }  
    }
}

var addOperators = function(num, target) {
    result = [];
    if(num === null || num.length === 0) return result;
    helper(num, 0, 0, 0, "", target);
    return result;
};